src/aes.rs

```diff
@@ -79,16 +79,61 @@ pub fn substitute_16_bytes(
     cs: ConstraintSystemRef<Fq>,
 ) -> (u128, ConstraintSystemRef<Fq>) {
     let num_witness =
-        UInt128::new_witness(ark_relations::ns!(cs, "substition_box_witness"), || Ok(num)).unwrap();
+        UInt128::new_witness(ark_relations::ns!(cs, "substitution_box_witness"), || {
+            Ok(num)
+        })
+        .unwrap();
     let mut new_bytes = [0u8; 16];
     let bytes = num_witness.to_bytes().unwrap();
     for i in 0..16 {
         new_bytes[i] = substitute_byte(bytes[i].value().unwrap());
     }
     return (u128::from_le_bytes(new_bytes), cs);
 }
+
+pub fn shift_rows(num: u128, cs: &ConstraintSystemRef<Fq>) -> u128 {
+    let num_witness =
+        UInt128::new_witness(ark_relations::ns!(cs, "shift_witness"), || Ok(num)).unwrap();
+
+    // Turn the 128 bit witness into
+    // its actual value, in the form of a
+    // vector of little endian bytes.
+    let mut witness_as_bytes = num_witness
+        .to_bytes()
+        .unwrap()
+        .into_iter()
+        .map(|byte| byte.value().unwrap())
+        .collect::<Vec<u8>>();
+
+    // Split the bytes as a Vec<&[u8]>, representing the AES state matrix.
+    let mut state_matrix = witness_as_bytes.chunks_mut(4).collect::<Vec<_>>();
+
+    // Rotate every state matrix row (u8 array) like specified by
+    // the AES cipher algorithm.
+    for (rotations, bytes) in state_matrix.iter_mut().enumerate() {
+        bytes.rotate_left(rotations);
+    }
+    // Turn the rotated arrays into a flattened
+    // 16 byte array, this is because the u128::from_le_bytes function
+    // only accepts 16 byte arrays.
+    let mut separated_bytes = [0u8; 16];
+    for (i, byte) in state_matrix.into_iter().flatten().enumerate() {
+        separated_bytes[i] = byte.clone();
```

Comment:
```suggestion
        separated_bytes[i] = byte;
```
This `clone` is not necessary because `u8` implements the `Copy` trait.

Reply:
Nice catch, I'll change it